File: java/src/main/java/com/squirrelparser/CSTNode.java
package com.squirrelparser;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.function.BiFunction;

/**
 * Base class for AST and CST nodes.
 */
abstract class Node<T extends Node<T>> {
    private final String label;
    private final int pos;
    private final int len;
    private final SyntaxError syntaxError;
    private final List<T> children;

    protected Node(String label, int pos, int len, SyntaxError syntaxError, List<T> children) {
        this.label = label;
        this.pos = pos;
        this.len = len;
        this.syntaxError = syntaxError;
        this.children = children;
    }

    public String label() { return label; }
    public int pos() { return pos; }
    public int len() { return len; }
    public SyntaxError syntaxError() { return syntaxError; }
    public List<T> children() { return children; }

    public String getInputSpan(String input) {
        return input.substring(pos, pos + len);
    }

    @Override
    public String toString() {
        return label + ": pos: " + pos + ", len: " + len;
    }

    public String toPrettyString(String input) {
        var buffer = new StringBuilder();
        buildTree(input, "", buffer, true);
        return buffer.toString();
    }

    protected void buildTree(String input, String prefix, StringBuilder buffer, boolean isRoot) {
        if (!isRoot) {
            buffer.append("\n");
        }
        buffer.append(prefix);
        buffer.append(label);
        if (children.isEmpty()) {
            buffer.append(": \"").append(getInputSpan(input)).append("\"");
        }

        for (int i = 0; i < children.size(); i++) {
            boolean isLast = i == children.size() - 1;
            String childPrefix = prefix + (isRoot ? "" : (isLast ? "    " : "|   "));
            String connector = isLast ? "`---" : "|---";

            buffer.append("\n");
            buffer.append(prefix);
            buffer.append(isRoot ? "" : connector);

            children.get(i).buildTree(input, childPrefix, buffer, false);
        }
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Build an AST from a parse tree.
 */
final class ASTBuilder {
    private ASTBuilder() {}

    public static ASTNode buildAST(ParseResult parseResult) {
        ASTNode extraNode = parseResult.unmatchedInput() != null
            ? ASTNode.syntaxError(parseResult.unmatchedInput())
            : null;

        return newASTNode(
            parseResult.topRuleName(),
            parseResult.root(),
            parseResult.transparentRules(),
            extraNode
        );
    }

    private static ASTNode newASTNode(String label, MatchResult refdMatchResult,
                                       Set<String> transparentRules, ASTNode addExtraASTNode) {
        List<ASTNode> childASTNodes = new ArrayList<>();
        collectChildASTNodes(refdMatchResult, childASTNodes, transparentRules);
        if (addExtraASTNode != null) {
            childASTNodes.add(addExtraASTNode);
        }
        return ASTNode.nonTerminal(label, childASTNodes);
    }

    private static void collectChildASTNodes(MatchResult matchResult,
                                              List<ASTNode> collectedAstNodes,
                                              Set<String> transparentRules) {
        if (matchResult.isMismatch()) {
            return;
        }
        if (matchResult instanceof SyntaxError se) {
            collectedAstNodes.add(ASTNode.syntaxError(se));
        } else {
            Clause clause = matchResult.clause();
            if (clause instanceof Terminal) {
                collectedAstNodes.add(ASTNode.terminal(matchResult));
            } else if (clause instanceof Ref ref) {
                if (!transparentRules.contains(ref.ruleName())) {
                    collectedAstNodes.add(newASTNode(
                        ref.ruleName(),
                        matchResult.subClauseMatches().getFirst(),
                        transparentRules,
                        null
                    ));
                }
            } else {
                for (MatchResult subClauseMatch : matchResult.subClauseMatches()) {
                    collectChildASTNodes(subClauseMatch, collectedAstNodes, transparentRules);
                }
            }
        }
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Base class for CST nodes.
 */
abstract class CSTNodeBase extends Node<CSTNodeBase> {
    protected CSTNodeBase(ASTNode astNode, List<CSTNodeBase> children) {
        super(astNode.label(), astNode.pos(), astNode.len(), astNode.syntaxError(), children);
    }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Factory function type for creating CST nodes from AST nodes.
 */
@FunctionalInterface
interface CSTNodeFactoryFn extends BiFunction<ASTNode, List<CSTNodeBase>, CSTNodeBase> {}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Build a CST from an AST.
 */
final class CSTBuilder {
    private CSTBuilder() {}

    /**
     * Build a CST from an AST using the provided factory map.
     *
     * <p>The factories map should contain an entry for each rule name in the grammar, plus:
     * <ul>
     *   <li>'&lt;Terminal&gt;' for terminal matches (string literals, character classes, etc.)</li>
     *   <li>'&lt;SyntaxError&gt;' if allowSyntaxErrors is true</li>
     * </ul>
     *
     * @param ast The AST to convert
     * @param factories Map from rule name to factory function
     * @param allowSyntaxErrors Whether to allow syntax errors
     * @return The CST root node
     */
    public static CSTNodeBase buildCST(
            ASTNode ast,
            java.util.Map<String, CSTNodeFactoryFn> factories,
            boolean allowSyntaxErrors) {
        return buildCSTInternal(ast, factories, allowSyntaxErrors);
    }

    private static CSTNodeBase buildCSTInternal(
            ASTNode ast,
            java.util.Map<String, CSTNodeFactoryFn> factories,
            boolean allowSyntaxErrors) {
        if (ast.syntaxError() != null) {
            if (!allowSyntaxErrors) {
                throw new IllegalArgumentException("Syntax error: " + ast.syntaxError());
            }
            var errorFactory = factories.get("<SyntaxError>");
            if (errorFactory == null) {
                throw new IllegalArgumentException("No factory found for <SyntaxError>");
            }
            return errorFactory.apply(ast, List.of());
        }

        var factory = factories.get(ast.label());

        if (factory == null && ast.label().equals(Terminal.NODE_LABEL)) {
            factory = factories.get("<Terminal>");
        }

        if (factory == null) {
            throw new IllegalArgumentException("No factory found for rule \"" + ast.label() + "\"");
        }

        List<CSTNodeBase> childCSTNodes = ast.children().stream()
            .map(child -> buildCSTInternal(child, factories, allowSyntaxErrors))
            .toList();
        return factory.apply(ast, childCSTNodes);
    }
}

File: python/squirrelparser/cst_node.py
"""AST and CST node classes."""

from __future__ import annotations
from abc import ABC
from typing import TYPE_CHECKING, Callable, Generic, TypeVar

from .match_result import SyntaxError, MatchResult
from .terminals import Terminal
from .combinators import Ref

if TYPE_CHECKING:
    from .parser import ParseResult

T = TypeVar('T', bound='Node')


# ------------------------------------------------------------------------------------------------------------------


class Node(ABC, Generic[T]):
    """Base class for AST and CST nodes."""

    __slots__ = ('label', 'pos', 'len', 'syntax_error', 'children')

    def __init__(
        self,
        *,
        label: str,
        pos: int,
        length: int,
        syntax_error: SyntaxError | None,
        children: list[T],
    ) -> None:
        self.label = label
        self.pos = pos
        self.len = length
        self.syntax_error = syntax_error
        self.children = children

    def get_input_span(self, input_str: str) -> str:
        return input_str[self.pos : self.pos + self.len]

    def __repr__(self) -> str:
        return f'{self.label}: pos: {self.pos}, len: {self.len}'

    def to_pretty_string(self, input_str: str) -> str:
        buffer: list[str] = []
        self._build_tree(input_str, '', buffer, True)
        return ''.join(buffer)

    def _build_tree(self, input_str: str, prefix: str, buffer: list[str], is_root: bool) -> None:
        if not is_root:
            buffer.append('\n')
        buffer.append(prefix)
        buffer.append(self.label)
        if not self.children:
            buffer.append(f': "{self.get_input_span(input_str)}"')

        for i, child in enumerate(self.children):
            is_last = i == len(self.children) - 1
            child_prefix = prefix + ('' if is_root else ('    ' if is_last else '|   '))
            connector = '`---' if is_last else '|---'

            buffer.append('\n')
            buffer.append(prefix)
            buffer.append('' if is_root else connector)

            child._build_tree(input_str, child_prefix, buffer, False)


# ------------------------------------------------------------------------------------------------------------------


class ASTNode(Node['ASTNode']):
    """An AST node representing either a rule match or a terminal match."""

    def __init__(
        self,
        *,
        label: str,
        pos: int,
        length: int,
        syntax_error: SyntaxError | None = None,
        children: list[ASTNode] | None = None,
    ) -> None:
        super().__init__(
            label=label,
            pos=pos,
            length=length,
            syntax_error=syntax_error,
            children=children if children is not None else [],
        )

    @classmethod
    def terminal(cls, terminal_match: MatchResult) -> ASTNode:
        return cls(
            label=Terminal.node_label,
            pos=terminal_match.pos,
            length=terminal_match.len,
        )

    @classmethod
    def non_terminal(cls, label: str, pos: int, length: int, children: list[ASTNode]) -> ASTNode:
        return cls(label=label, pos=pos, length=length, children=children)

    @classmethod
    def syntax_error_node(cls, se: SyntaxError) -> ASTNode:
        return cls(
            label=SyntaxError.node_label,
            pos=se.pos,
            length=se.len,
            syntax_error=se,
        )


# ------------------------------------------------------------------------------------------------------------------


def build_ast(parse_result: ParseResult) -> ASTNode:
    """Build an AST from a parse tree."""
    extra_node = (
        ASTNode.syntax_error_node(parse_result.unmatched_input)
        if parse_result.unmatched_input is not None
        else None
    )

    return _new_ast_node(
        parse_result.top_rule_name,
        parse_result.root,
        parse_result.transparent_rules,
        extra_node,
    )


def _new_ast_node(
    label: str,
    refd_match_result: MatchResult,
    transparent_rules: set[str],
    add_extra_ast_node: ASTNode | None,
) -> ASTNode:
    child_ast_nodes: list[ASTNode] = []
    _collect_child_ast_nodes(refd_match_result, child_ast_nodes, transparent_rules)
    if add_extra_ast_node is not None:
        child_ast_nodes.append(add_extra_ast_node)
    return ASTNode.non_terminal(label, refd_match_result.pos, refd_match_result.len, child_ast_nodes)


def _collect_child_ast_nodes(
    match_result: MatchResult,
    collected_ast_nodes: list[ASTNode],
    transparent_rules: set[str],
) -> None:
    if match_result.is_mismatch:
        return
    if isinstance(match_result, SyntaxError):
        collected_ast_nodes.append(ASTNode.syntax_error_node(match_result))
    else:
        clause = match_result.clause
        if isinstance(clause, Terminal):
            collected_ast_nodes.append(ASTNode.terminal(match_result))
        elif isinstance(clause, Ref):
            if clause.rule_name not in transparent_rules:
                collected_ast_nodes.append(
                    _new_ast_node(
                        clause.rule_name,
                        match_result.sub_clause_matches[0],
                        transparent_rules,
                        None,
                    )
                )
        else:
            for sub_clause_match in match_result.sub_clause_matches:
                _collect_child_ast_nodes(sub_clause_match, collected_ast_nodes, transparent_rules)


# ------------------------------------------------------------------------------------------------------------------


class CSTNode(Node['CSTNode'], ABC):
    """Base class for CST nodes."""

    def __init__(self, ast_node: ASTNode, children: list[CSTNode]) -> None:
        super().__init__(
            label=ast_node.label,
            pos=ast_node.pos,
            length=ast_node.len,
            syntax_error=ast_node.syntax_error,
            children=children,
        )


# ------------------------------------------------------------------------------------------------------------------


# Type alias for factory functions that create CST nodes from AST nodes
CSTNodeFactoryFn = Callable[[ASTNode, list['CSTNode']], 'CSTNode']


# ------------------------------------------------------------------------------------------------------------------


def build_cst(
    ast: ASTNode,
    factories: dict[str, CSTNodeFactoryFn],
    allow_syntax_errors: bool,
) -> CSTNode:
    """
    Build a CST from an AST using the provided factory map.

    The factories map should contain an entry for each rule name in the grammar, plus:
    - '<Terminal>' for terminal matches (string literals, character classes, etc.)
    - '<SyntaxError>' if allow_syntax_errors is True
    """
    return _build_cst_internal(ast, factories, allow_syntax_errors)


def _build_cst_internal(
    ast: ASTNode,
    factories: dict[str, CSTNodeFactoryFn],
    allow_syntax_errors: bool,
) -> CSTNode:
    if ast.syntax_error is not None:
        if not allow_syntax_errors:
            raise ValueError(f'Syntax error: {ast.syntax_error}')
        error_factory = factories.get('<SyntaxError>')
        if error_factory is None:
            raise ValueError('No factory found for <SyntaxError>')
        return error_factory(ast, [])

    factory = factories.get(ast.label)

    if factory is None and ast.label == Terminal.node_label:
        factory = factories.get('<Terminal>')

    if factory is None:
        raise ValueError(f'No factory found for rule "{ast.label}"')

    child_cst_nodes = [_build_cst_internal(child, factories, allow_syntax_errors) for child in ast.children]
    return factory(ast, child_cst_nodes)
